### PR TITLE
Deploying operator and the driver in non-default namespace

### LINKED ISSUE #785: pmem-csi-operator installation in a separate non-"default" namespace
Current pmem-csi-operator manifest ([https://github.com/intel/pmem-csi/raw/v0.8.0/deploy/operator/pmem-csi-operator.yaml](url)) uses default namespace for installation. 
It is common knowledge and best practice to not use namespace "default" for anything. The operator should use dedicated namespace named "pmem-csi-operator" (or similar) and documentation should provide instruction how to deploy it in different namespace (if necessary).

### PR DESCRIPTION
**Pull request by avalluri**
Deployment changes to run the operator and the driver in the non-default namespace. By default, the manifests deploy in 'pmem-csi' namespace. Also had to fix tests to support this change.
FIXES #785

### CODE REVIEW

File: docs/install.md [hunk 1]
```diff
@@ -263,6 +263,7 @@ $ kubectl create -f https://github.com/intel/pmem-csi/raw/devel/deploy/crd/pmem-
 ``` console
 $ kubectl create -f https://github.com/intel/pmem-csi/raw/devel/deploy/operator/pmem-csi-operator.yaml
 ```
+The operator gets deployed in 'pmem-csi' namespace.
```
Comment (pohly) [inline docs/install.md:266]: "in a namespace called 'pmem-csi'" which gets created by that YAML file"

File: docs/install.md [hunk 2]
```diff
@@ -367,7 +368,7 @@ Events:
   Normal  NewDeployment  58s   pmem-csi-operator  Processing new driver deployment
   Normal  Running        39s   pmem-csi-operator  Driver deployment successful
 
-$ kubectl get po
+$ kubectl get po -n pmem-csi
```
Comment (pohly) [inline docs/install.md:371]: While at it, can you replace `po` with `pods`?

**Review by pohly — CHANGES_REQUESTED** [review]

**Comment by pohly** [comment]
"make test" failed. Please, remember to run that before submitting a change.

File: docs/install.md (hunk 2 above)
Comment (avalluri) [reply docs/install.md:371]: Done.

File: docs/install.md (hunk 1 above)
Comment (avalluri) [reply docs/install.md:266]: Changed the text as suggested.

File: deploy/kustomize/operator/kustomization.yaml [hunk 3]
```diff
@@ -1,2 +1,34 @@
...
+  objref:
+    apiVersion: v1
+    kind: ServiceAccount
+    name: pmem-csi-operator
+  fieldRef:
+    fieldPath: metadata.namespace
+
+configurations:
+- kustomizeconfig.yaml
+
```
[the remark this answers is not in the document]
Comment (avalluri) [reply deploy/kustomize/operator/kustomization.yaml:34]: Done.

File: deploy/kustomize/operator/namespace.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: default
```
[the remark this answers is not in the document]
Comment (avalluri) [reply deploy/kustomize/operator/namespace.yaml]: Done.

File: deploy/kustomize/operator/kustomizeconfig.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+varReference:
+- kind: RoleBinding
+  group: rbac.authorization.k8s.io
+  path: subjects/namespace
```
[the remark this answers is not in the document]
Comment (avalluri) [reply deploy/kustomize/operator/kustomizeconfig.yaml:4]: Done.

File: deploy/kustomize/operator/events-rbac.yaml [hunk 6]
```diff
@@ -0,0 +1,29 @@
+#
+# RBAC rules required for the operator to generate
+# deployment events. Events are generated in the default
+# namespace even if the operator deployed in different namespace
+apiVersion: rbac.authorization.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (avalluri) [reply deploy/kustomize/operator/events-rbac.yaml:5]: Done.

**Comment by avalluri** [comment]
> "make test" failed. Please, remember to run that before submitting a change.

I did run the `make test` and other operator tests quite enough times. But it didn't occur on my setup. It is only visible on a clean clone :(. Hopefully, I fixed it now.

**Comment by pohly** [comment]
>  It is only visible on a clean clone :(

Why was that? That sounds like an issue in our testing process that we may want to fix.

**Comment by pohly** [comment]
Something broke consistently in "switch device mode lvm-to-direct delete volume – fedora-1_17.operator.API": https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-829/2/tests

```
Dec 8 22:47:12.557: Giving up waiting for PMEM-CSI to start up, check the previous warnings and log output. Last error: get port for service lvm-to-direct-delete-volume-metrics in namespace default: services "lvm-to-direct-delete-volume-metrics" not found
```

**Comment by pohly** [comment]
Do we still get our pod output dumped when running E2E tests? Probably not, because https://github.com/intel/pmem-csi/blob/3e5744b93675b0b7aed231de8eb8b57bc1ca2454/test/e2e/e2e.go#L111-L121 is only watching the "default" namespace.

That should be changed, ideally so that it works for all test scenarios. We were missing the operator output already before when deployed in kube-system (a special case). That couldn't be avoided because watching everything in "kube-system" was too verbose (I tried), but at least "pmem-csi" and whatever other PMEM-CSI specific namespaces are used now should be doable.

**Comment by avalluri** [comment]
> Something broke consistently in "switch device mode lvm-to-direct delete volume – fedora-1_17.operator.API": https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-829/2/tests
> 
> ```
> Dec 8 22:47:12.557: Giving up waiting for PMEM-CSI to start up, check the previous warnings and log output. Last error: get port for service lvm-to-direct-delete-volume-metrics in namespace default: services "lvm-to-direct-delete-volume-metrics" not found
> ```

The issue is with the test itself it was waiting for the driver pods in the `default` namespace whereas the operator deployed in `operator-test` namespace.

**Comment by pohly** [comment]
I'm wondering whether version skew tests still work with this change. Have you tested that manually? Those tests don't run in pre-submit jobs.

**Comment by avalluri** [comment]
> is only watching the "default" namespace.
> That should be changed, ideally so that it works for all test scenarios. We were missing the operator output already before when deployed in kube-system (a special case). That couldn't be avoided because watching everything in "kube-system" was too verbose (I tried), but at least "pmem-csi" and whatever other PMEM-CSI specific namespaces are used now should be doable.

Added the "operator-test" namespace to the watchlist, this is the namespace used by the operator tests.

**Comment by avalluri** [comment]
> I'm wondering whether version skew tests still work with this change. Have you tested that manually? Those tests don't run in pre-submit jobs.

@pohly How I can run them locally?

**Comment by pohly** [comment]
> How I can run them locally?

`make test_e2e TEST_E2E_FOCUS=skew`

**Comment by avalluri** [comment]
> I'm wondering whether version skew tests still work with this change. Have you tested that manually? Those tests don't run in pre-submit jobs.

I tested version skew tests with this change locally. There was one issue discover about the driver certificates. In this change by default certificates are generated for the 'pmem-csi' namespace which are not usable for the driver running in default namespace.  I fixed this.

Except that I couldn't figure out any other issues.

**Review by pohly — APPROVED** [review]

**Pull request merged**